cbt_export_changes.py

```diff
@@ -24,37 +24,36 @@
 from bitstring import BitStream
 import argparse
 import os
+import re
 
 # CBT tracks 64KB blocks. Therefore each bit in the bitmap corresponds to a
 # 64KB block on the VDI.
 changed_block_size = 64 * 1024
 certfile = "cacert.pem"
 
 
-def collect_host_certificates(session):
-    # Note that this function is only necessary for self-signed certificates
-    # For certificates signed by a CA you will need to collect the public
-    # certificates from the CA.
-    host_list = session.xenapi.host.get_all()
-    with open(certfile, 'w') as cert_out:
-        for host in host_list:
-            cert_out.write(session.xenapi.host.get_server_certificate(host)
-                           + "\n")
+def find_fqdn(cert_text):
+    from cryptography import x509
+    from cryptography.hazmat.backends import default_backend
+    cert = x509.load_pem_x509_certificate(cert_text, default_backend())
+    cert_subject = str(cert.subject)
+    cert_hostname = re.search(r'value=u\'(.*)\'', cert_subject)

```

Comment:
I tried multiple methods of getting the hostname from the cert without having to convert it to a string and search it but could not get it to work. The cert is an object type, doesn't seem to have any attribute that I could work out that just gives you the hostname.

Reply:
Do we expect this to work for the different certificate formats, as indicated by Thomas?

Reply:
I have only tested using the basic autogenerated certificate however it looks like the cryptography python library has some inbuilt logic to deal with the hostname being populated in different places which I think should mean it will work for a range of different certificate formats.

Reply:
@robertbreker The certificate file on the server (and hence the `cert_text` here) will always be in PEM format (which is just base64-encoded DER format plus a header and footer).

@thomasmck If you're using the `cryptography` module I think you are, you should base this on the example at
https://cryptography.io/en/latest/x509/reference/#cryptography.x509.SubjectAlternativeName
and only use `cert.subject` if there is no subject alternative name.

As for getting from one of those to a subject string, see
https://cryptography.io/en/latest/x509/reference/#cryptography.x509.DNSName
https://cryptography.io/en/latest/x509/reference/#cryptography.x509.GeneralName

If you have to fall back to `cert.subject`, that's an instance of
https://cryptography.io/en/latest/x509/reference/#cryptography.x509.Name
from https://cryptography.io/en/latest/x509/reference/#cryptography.x509.Certificate.subject